{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      Genome  Genome ID  \\\n",
      "0  Escherichia coli str. K-12 substr. MG1655  511145.12   \n",
      "1      Escherichia coli O83:H1 str. NRG 857C   685038.3   \n",
      "2                     Escherichia coli IAI39   585057.6   \n",
      "3                    Escherichia coli UMN026   585056.7   \n",
      "4           Escherichia coli strain NCTC9001   562.7382   \n",
      "\n",
      "                PATRIC ID RefSeq Locus Tag   Gene                     Product  \\\n",
      "0  fig|511145.12.peg.3613            b3506    slp  Starvation lipoprotein Slp   \n",
      "1   fig|685038.3.peg.3547     NRG857_17375    NaN  Starvation lipoprotein Slp   \n",
      "2   fig|585057.6.peg.4139     ECIAI39_3999    slp  Starvation lipoprotein Slp   \n",
      "3   fig|585056.7.peg.4170       ECUMN_3997    slp  Starvation lipoprotein Slp   \n",
      "4    fig|562.7382.peg.510  ERS451419_00510  slp_1  Starvation lipoprotein Slp   \n",
      "\n",
      "  Length (NT)  Length (AA)  ALN Length  Identity  Query cover  Subject cover  \\\n",
      "0         567          188         188        98          100            100   \n",
      "1         567          188         188        98          100            100   \n",
      "2         567          188         188        98          100            100   \n",
      "3         567          188         188        98          100            100   \n",
      "4         567          188         188        98          100            100   \n",
      "\n",
      "   Hit from  Hit to  Score        E value  Unnamed: 16  \n",
      "0         1     188    374  4.000000e-130          NaN  \n",
      "1         1     188    374  9.000000e-130          NaN  \n",
      "2         1     188    374  9.000000e-130          NaN  \n",
      "3         1     188    374  9.000000e-130          NaN  \n",
      "4         1     188    374  9.000000e-130          NaN  \n",
      "                PATRIC ID                                          Sequences\n",
      "0  fig|1004326.3.peg.3129  MSLFSPKIFTIALFSVFLTACSSLPENLDTTNESVVTDYQLWVNMK...\n",
      "1  fig|1005043.3.peg.1286  MKIEIFSAKFFFAVLATLLLTGCVNIPKEIQGKTETPQQNLNLIKN...\n",
      "2   fig|1001530.3.peg.658  MFRKLFLGLILIFMVGCASVPASLQTTTKNPITDLHVITEHPTVMQ...\n",
      "3  fig|1002368.3.peg.2942  MQVVRGLLAGIIVAALSGCVSIPDAIKGSSPTPQQDLVRVMAAPQL...\n",
      "4  fig|1005995.4.peg.2288  MMKLIKHCPAMLLVPVLLLSGCVTVPQSIRGHSPLPQQNLVQVMNA...\n"
     ]
    }
   ],
   "source": [
    "#Read your blast results csv in\n",
    "df = pd.read_csv(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\slp_blast.csv\")\n",
    "#Show the first 5 rows\n",
    "print(df.head())\n",
    "#Temp lists to attatch variables \n",
    "ids = []\n",
    "seqs = []\n",
    "#Seq.IO.parse is. an easy way to read through every gene/prot in a fasta file \n",
    "for record in SeqIO.parse(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\slp.fasta\", \"fasta\"):\n",
    "    # So the fasta ID line is fig|1240239.3.peg.2328|XFF4834R_chr22840|VBIXanFus78442_2328|\n",
    "    # but we only need the first two elements (\"fig and 1240239.3.peg.2328\") to make fig|1240239.3.peg.2328\n",
    "    # We then use this ID to merge back onto the original dataframe (basically we say if this row has this ID, add the sequence to that row)\n",
    "    \n",
    "    #Fancy way of extracting the ID\n",
    "    ids.append(\"|\".join(record.id.split(\"|\")[0:2]))\n",
    "    #Fancy way of extracting the sequence \n",
    "    seqs.append(str(record.seq))\n",
    "#making a new dataframe from this info \n",
    "seq_df = pd.DataFrame.from_dict({\"PATRIC ID\": ids, \"Sequences\":seqs})\n",
    "print(seq_df.head())\n",
    "#Mergine back and saving \n",
    "df = df.merge(seq_df, on=\"PATRIC ID\")\n",
    "df.to_csv(\"/Users/cmgro/Downloads/merged_slp.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 499 entries, 0 to 498\n",
      "Data columns (total 19 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   Unnamed: 0        499 non-null    int64  \n",
      " 1   Genome            499 non-null    object \n",
      " 2   Genome ID         499 non-null    float64\n",
      " 3   PATRIC ID         499 non-null    object \n",
      " 4   RefSeq Locus Tag  364 non-null    object \n",
      " 5   Gene              58 non-null     object \n",
      " 6   Product           499 non-null    object \n",
      " 7   Length (NT)       499 non-null    int64  \n",
      " 8   Length (AA)       499 non-null    int64  \n",
      " 9   ALN Length        499 non-null    int64  \n",
      " 10  Identity          499 non-null    int64  \n",
      " 11  Query cover       499 non-null    int64  \n",
      " 12  Subject cover     499 non-null    int64  \n",
      " 13  Hit from          499 non-null    int64  \n",
      " 14  Hit to            499 non-null    int64  \n",
      " 15  Score             499 non-null    int64  \n",
      " 16  E value           499 non-null    float64\n",
      " 17  Unnamed: 16       0 non-null      float64\n",
      " 18  Sequences         499 non-null    object \n",
      "dtypes: float64(3), int64(10), object(6)\n",
      "memory usage: 74.2+ KB\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 20 entries, 0 to 19\n",
      "Data columns (total 19 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   Unnamed: 0        20 non-null     int64  \n",
      " 1   Genome            20 non-null     object \n",
      " 2   Genome ID         20 non-null     float64\n",
      " 3   PATRIC ID         20 non-null     object \n",
      " 4   RefSeq Locus Tag  20 non-null     object \n",
      " 5   Gene              10 non-null     object \n",
      " 6   Product           20 non-null     object \n",
      " 7   Length (NT)       20 non-null     int64  \n",
      " 8   Length (AA)       20 non-null     int64  \n",
      " 9   ALN Length        20 non-null     int64  \n",
      " 10  Identity          20 non-null     int64  \n",
      " 11  Query cover       20 non-null     int64  \n",
      " 12  Subject cover     20 non-null     int64  \n",
      " 13  Hit from          20 non-null     int64  \n",
      " 14  Hit to            20 non-null     int64  \n",
      " 15  Score             20 non-null     int64  \n",
      " 16  E value           20 non-null     float64\n",
      " 17  Unnamed: 16       0 non-null      float64\n",
      " 18  Sequences         20 non-null     object \n",
      "dtypes: float64(3), int64(10), object(6)\n",
      "memory usage: 3.1+ KB\n",
      "None\n",
      "{'Starvation lipoprotein Slp': 23, 'hypothetical protein': 94, 'Uncharacterized lipoprotein YeaY': 176, 'Starvation lipoprotein Slp paralog': 201, 'outer membrane lipoprotein, Slp family': 1, 'outer membrane lipoprotein Slp': 3, 'outer membrane protein slp precursor': 1}\n",
      "{'Starvation lipoprotein Slp': 20}\n"
     ]
    }
   ],
   "source": [
    "# reads blast results\n",
    "df = pd.read_csv(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\merged_slp.csv\")\n",
    "\n",
    "print(df.info())\n",
    "\n",
    "# filters out entries with query cover and identity values less than 50%\n",
    "df_filtered = df.loc[(df['Query cover']>=50) & (df['Identity']>=50)]\n",
    "\n",
    "# prints info on the new dataframe \n",
    "print(df_filtered.info())\n",
    "\n",
    "def count_types(types):\n",
    "    types_counts = {}\n",
    "    for i in types:\n",
    "        if i in types_counts.keys():\n",
    "            types_counts[i] += 1\n",
    "        else: \n",
    "            types_counts[i] = 1\n",
    "    return types_counts\n",
    "print(count_types(df[\"Product\"]))\n",
    "print(count_types(df_filtered[\"Product\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
