{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      Genome  Genome ID  \\\n",
      "0  Escherichia coli str. K-12 substr. MG1655  511145.12   \n",
      "1      Escherichia coli O83:H1 str. NRG 857C   685038.3   \n",
      "2                     Escherichia coli IAI39   585057.6   \n",
      "3                    Escherichia coli UMN026   585056.7   \n",
      "4           Escherichia coli strain NCTC9001   562.7382   \n",
      "\n",
      "                PATRIC ID RefSeq Locus Tag   Gene                     Product  \\\n",
      "0  fig|511145.12.peg.3613            b3506    slp  Starvation lipoprotein Slp   \n",
      "1   fig|685038.3.peg.3547     NRG857_17375    NaN  Starvation lipoprotein Slp   \n",
      "2   fig|585057.6.peg.4139     ECIAI39_3999    slp  Starvation lipoprotein Slp   \n",
      "3   fig|585056.7.peg.4170       ECUMN_3997    slp  Starvation lipoprotein Slp   \n",
      "4    fig|562.7382.peg.510  ERS451419_00510  slp_1  Starvation lipoprotein Slp   \n",
      "\n",
      "  Length (NT)  Length (AA)  ALN Length  Identity  Query cover  Subject cover  \\\n",
      "0         567          188         188        98          100            100   \n",
      "1         567          188         188        98          100            100   \n",
      "2         567          188         188        98          100            100   \n",
      "3         567          188         188        98          100            100   \n",
      "4         567          188         188        98          100            100   \n",
      "\n",
      "   Hit from  Hit to  Score        E value  Unnamed: 16  \n",
      "0         1     188    374  4.000000e-130          NaN  \n",
      "1         1     188    374  9.000000e-130          NaN  \n",
      "2         1     188    374  9.000000e-130          NaN  \n",
      "3         1     188    374  9.000000e-130          NaN  \n",
      "4         1     188    374  9.000000e-130          NaN  \n",
      "                PATRIC ID                                          Sequences\n",
      "0   fig|1001530.3.peg.839  MKYKRILSQSHVVLKEFYVDHHLIVKLSNVQGDVHINGHAIELKNN...\n",
      "1  fig|1002368.3.peg.2677  MTEKDVDVKINDTTSLILPANHLNVISCNNNIIDFSYLDNALVAHM...\n",
      "2   fig|1003237.3.peg.556  MANALIIRNEDLDKPLVSVLSPEDEKLKSVLERHLLDLMGPAAAPG...\n",
      "3  fig|1002368.3.peg.2661  MQNVQLQHKHLTAAEVTSGQMHRLHRVKLFSAAICHITQGSKVIIQ...\n",
      "4  fig|1006551.4.peg.3266  MRNIHFHHNALTAAEITTRRLCKLHRVTLFSPAICRVNRGSKVIVQ...\n"
     ]
    }
   ],
   "source": [
    "#Read your blast results csv in\n",
    "df = pd.read_csv(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\slp_blast.csv\")\n",
    "#Show the first 5 rows\n",
    "print(df.head())\n",
    "#Temp lists to attatch variables \n",
    "ids = []\n",
    "seqs = []\n",
    "#Seq.IO.parse is. an easy way to read through every gene/prot in a fasta file \n",
    "for record in SeqIO.parse(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\slp_genome_feature (6).fasta\", \"fasta\"):\n",
    "    # So the fasta ID line is fig|1240239.3.peg.2328|XFF4834R_chr22840|VBIXanFus78442_2328|\n",
    "    # but we only need the first two elements (\"fig and 1240239.3.peg.2328\") to make fig|1240239.3.peg.2328\n",
    "    # We then use this ID to merge back onto the original dataframe (basically we say if this row has this ID, add the sequence to that row)\n",
    "    \n",
    "    #Fancy way of extracting the ID\n",
    "    ids.append(\"|\".join(record.id.split(\"|\")[0:2]))\n",
    "    #Fancy way of extracting the sequence \n",
    "    seqs.append(str(record.seq))\n",
    "#making a new dataframe from this info \n",
    "seq_df = pd.DataFrame.from_dict({\"PATRIC ID\": ids, \"Sequences\":seqs})\n",
    "print(seq_df.head())\n",
    "#Mergine back and saving \n",
    "df = df.merge(seq_df, on=\"PATRIC ID\")\n",
    "df.to_csv(\"/Users/cmgro/Downloads/merged_slp.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 0 entries\n",
      "Data columns (total 19 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   Unnamed: 0        0 non-null      object\n",
      " 1   Genome            0 non-null      object\n",
      " 2   Genome ID         0 non-null      object\n",
      " 3   PATRIC ID         0 non-null      object\n",
      " 4   RefSeq Locus Tag  0 non-null      object\n",
      " 5   Gene              0 non-null      object\n",
      " 6   Product           0 non-null      object\n",
      " 7   Length (NT)       0 non-null      object\n",
      " 8   Length (AA)       0 non-null      object\n",
      " 9   ALN Length        0 non-null      object\n",
      " 10  Identity          0 non-null      object\n",
      " 11  Query cover       0 non-null      object\n",
      " 12  Subject cover     0 non-null      object\n",
      " 13  Hit from          0 non-null      object\n",
      " 14  Hit to            0 non-null      object\n",
      " 15  Score             0 non-null      object\n",
      " 16  E value           0 non-null      object\n",
      " 17  Unnamed: 16       0 non-null      object\n",
      " 18  Sequences         0 non-null      object\n",
      "dtypes: object(19)\n",
      "memory usage: 0.0+ bytes\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 0 entries\n",
      "Data columns (total 19 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   Unnamed: 0        0 non-null      object\n",
      " 1   Genome            0 non-null      object\n",
      " 2   Genome ID         0 non-null      object\n",
      " 3   PATRIC ID         0 non-null      object\n",
      " 4   RefSeq Locus Tag  0 non-null      object\n",
      " 5   Gene              0 non-null      object\n",
      " 6   Product           0 non-null      object\n",
      " 7   Length (NT)       0 non-null      object\n",
      " 8   Length (AA)       0 non-null      object\n",
      " 9   ALN Length        0 non-null      object\n",
      " 10  Identity          0 non-null      object\n",
      " 11  Query cover       0 non-null      object\n",
      " 12  Subject cover     0 non-null      object\n",
      " 13  Hit from          0 non-null      object\n",
      " 14  Hit to            0 non-null      object\n",
      " 15  Score             0 non-null      object\n",
      " 16  E value           0 non-null      object\n",
      " 17  Unnamed: 16       0 non-null      object\n",
      " 18  Sequences         0 non-null      object\n",
      "dtypes: object(19)\n",
      "memory usage: 0.0+ bytes\n",
      "None\n",
      "{}\n",
      "{}\n"
     ]
    }
   ],
   "source": [
    "# reads blast results\n",
    "df = pd.read_csv(\"C:\\\\Users\\\\cmgro\\\\Downloads\\\\merged_slp.csv\")\n",
    "\n",
    "print(df.info())\n",
    "\n",
    "# filters out entries with query cover and identity values less than 50%\n",
    "df_filtered = df.loc[(df['Query cover']>=50) & (df['Identity']>=50)]\n",
    "\n",
    "# prints info on the new dataframe \n",
    "print(df_filtered.info())\n",
    "\n",
    "def count_types(types):\n",
    "    types_counts = {}\n",
    "    for i in types:\n",
    "        if i in types_counts.keys():\n",
    "            types_counts[i] += 1\n",
    "        else: \n",
    "            types_counts[i] = 1\n",
    "    return types_counts\n",
    "print(count_types(df[\"Product\"]))\n",
    "print(count_types(df_filtered[\"Product\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
